### PR TITLE
Add notebooks for first-hit analysis on aggregated levels and scripts for single-cell analysis.  

### PR DESCRIPTION
**Pull request by Arkkienkeli**
For single-cells, in case of TA-ORF there is a notebook, for BBBC022 a script (to run it in tmux as it takes time to run).

### CODE REVIEW

File: bbbc022/first-hits-single-cell.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+cols = ["Metadata_Plate","Metadata_Well","Metadata_Site","Plate_Map_Name","broad_sample_Replicate","Treatment","Compound","Concentration","Replicate_Use"]
+sc_meta = pd.merge(pd.DataFrame(sc_meta_idx, columns=["ID"]), meta[cols], left_on="ID", right_index=True)
+sc_meta = pd.merge(pd.DataFrame(sc_sc_idx, columns=["SCID"]), sc_meta, left_index=True, right_index=True)
+
+sc_controls = sc_meta[sc_meta[PERT_NAME] == CTRL_NAME]
+sc_treatments = sc_meta[sc_meta[PERT_NAME] != CTRL_NAME]
+
+feats = [i for i in range(NUM_FEATURES)]
+
+def request_controls(control_meta_subset):
```
Comment (jccaicedo) [inline bbbc022/first-hits-single-cell.py:79]: Move function up before the executable code starts (before line 80).

File: bbbc022/first-hits-single-cell.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        )
+        if os.path.isfile(filename):
+            with open(filename, "rb") as data:
+                info = np.load(data)
+                sc_control_features_array[i,:] = info["features"][control_meta_subset.loc[i, "SCID"]]
+    
+    return pd.merge(control_meta_subset, pd.DataFrame(data=sc_control_features_array), how="left", left_index=True, right_index=True)
+
+
+def request_treatments(treatments_meta_subset):
```
Comment (jccaicedo) [inline bbbc022/first-hits-single-cell.py:98]: Same here, move function up.

File: bbbc022/first-hits-single-cell.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        )
+        if os.path.isfile(filename):
+            with open(filename, "rb") as data:
+                info = np.load(data)
+                sc_treatment_features_array[i,:] = info["features"][treatments_meta_subset.loc[i, "SCID"]]
+    
+    return pd.merge(treatments_meta_subset, pd.DataFrame(data=sc_treatment_features_array), how="left", left_index=True, right_index=True)
+
+
+for i in range(10):
```
Comment (jccaicedo) [inline bbbc022/first-hits-single-cell.py]: Bring all the main code to this block and wrap it in the following conditional:
```
if __name__ == "__main__":
    # Executable code here
```
See why here: https://stackoverflow.com/questions/419163/what-does-if-name-main-do

File: bbbc022/first-hits-single-cell.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+REG_PARAM = 1e-2
+
+
+def find_first_hits(features, feats, pert_name, control):
+    results = []
+
+    gen = features[features[pert_name] != control].groupby(["Metadata_Plate", "Metadata_Well", pert_name])["Replicate_Use"].count().reset_index().iterrows()
+    for k, r in tqdm(gen):
+        # Select samples in a well
+        s = time.time()
```
Comment (jccaicedo) [inline bbbc022/first-hits-single-cell.py]: Can be removed

File: bbbc022/first-hits-single-cell.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
+import pandas as pd
+import numpy as np
+import seaborn as sb
+import os
+import sys
+import time
```
Comment (jccaicedo) [inline bbbc022/first-hits-single-cell.py]: Can be removed (unused in the logic)

File: bbbc022/first-hits-aggregated.ipynb [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "def find_first_hits(features, feats, pert_name, control):\n",
```
Comment (jccaicedo) [inline bbbc022/first-hits-aggregated.ipynb:62]: Is this function different for aggregated profiles and for single cells?
In principle, they are the same. We need to unify them and have only one implementation. Let's refactor and move these functions out to a module called `first_hits.py` in the `profiling` directory. In that way, we can reuse it in all other scripts.

File: bbbc022/first-hits-aggregated.ipynb [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "def summarize(results):\n",
```
Comment (jccaicedo) [inline bbbc022/first-hits-aggregated.ipynb:107]: Try to refactor and keep in a separate module that can be imported.

**Review by jccaicedo — CHANGES_REQUESTED** [review]

**Comment by jccaicedo** [comment]
The logic seems to be OK overall, but the scripts have redundant code. Let's make an effort to factorize and reuse code as much as we can across all scripts (especially functions). Ideally, each script in this PR should have logic that is specific to the dataset only.

**Comment by Arkkienkeli** [comment]
We eventually did not use this analysis in the final publication, thus we are not going to merge this code.

**Pull request closed without merging**